### PR TITLE
[InstrProf] Add debuginfod correlation support

### PR DESCRIPTION
**Pull request by gulfemsavrun**
This patch adds debuginfod support into llvm-profdata to
find the assosicated executable by a build id in a raw
profile to correlate a profile with a provided correlation
kind (debug-info or binary).

### CODE REVIEW

File: llvm/lib/ProfileData/InstrProfCorrelator.cpp [hunk 1]
```diff
@@ -113,11 +115,36 @@ InstrProfCorrelator::get(StringRef Filename, ProfCorrelatorKind FileKind) {
...
-    auto BufferOrErr = errorOrToExpected(MemoryBuffer::getFile(Filename));
-    if (auto Err = BufferOrErr.takeError())
-      return std::move(Err);
+    if (!Filename.empty()) {
+      auto BufferOrErr = errorOrToExpected(MemoryBuffer::getFile(Filename));
+      if (auto Err = BufferOrErr.takeError())
+        return std::move(Err);
+      return get(std::move(*BufferOrErr), FileKind);
+    } else if (BIDFetcher) {
+      if (BIs->size() > 1)
```
Comment (ZequanWu) [inline llvm/lib/ProfileData/InstrProfCorrelator.cpp:124]: Do you plan to add support for multiple build IDs in a single raw profile in the future? Like a lightweight raw profile contains two build ids, one for main executable and another one for a shared library. Ideally, BIDFetcher should find/download both the executable and shared library and the correlator will match them based on their build IDs.

File: llvm/lib/ProfileData/InstrProfCorrelator.cpp [hunk 2]
```diff
@@ -113,11 +115,36 @@ InstrProfCorrelator::get(StringRef Filename, ProfCorrelatorKind FileKind) {
...
+        return make_error<InstrProfError>(
+            instrprof_error::unable_to_correlate_profile,
+            "unsupported profile binary correlation when there are multiple "
+            "build IDs in a binary");
```
Comment (ZequanWu) [inline llvm/lib/ProfileData/InstrProfCorrelator.cpp:128]: IIUC, `BIs` is a vector of build IDs read from a raw profile. "in a binary" -> "in a raw profile".

**Review by ZequanWu — COMMENTED** [review]
1. Does the debuginfod correlation support work with both debug info and binary correlation? Looks like it only works with binary correlation in this change. I guess it won't be difficult to make it work with debug info correlation as well.
2. How does the new flag  `--debug-file-directory` interact with existing flag `--binary-file` which specifies just one unstripped binary?

File: llvm/lib/ProfileData/InstrProfCorrelator.cpp (hunk 1 above)
Comment (mysterymath) [inline llvm/lib/ProfileData/InstrProfCorrelator.cpp:124]: This code doesn't make a distinction between a `BIs` that is set to an empty array and a `BIs` that is unset. Indeed, the former case appears to crash. Accordingly, `BIs` should just be an `ArrayRef`.

File: llvm/include/llvm/ProfileData/InstrProfReader.h [hunk 3]
```diff
@@ -314,6 +316,11 @@ class RawInstrProfReader : public InstrProfReader {
   /// If available, this hold the ProfileData array used to correlate raw
   /// instrumentation data to their functions.
   const InstrProfCorrelatorImpl<IntPtrT> *Correlator;
+  /// Correlator that fetches debuginfo from debuginfod on the fly by build id.
```
Comment (mysterymath) [inline llvm/include/llvm/ProfileData/InstrProfReader.h:319]: nit: build ID

File: llvm/tools/llvm-profdata/llvm-profdata.cpp [hunk 4]
```diff
@@ -914,9 +922,14 @@ static void mergeInstrProfile(const WeightedFileVector &Inputs,
     exitWithError("unknown format is specified");
 
   // TODO: Maybe we should support correlation with mixture of different
-  // correlation modes(w/wo debug-info/object correlation).
-  if (!DebugInfoFilename.empty() && !BinaryFilename.empty())
-    exitWithError("Expected only one of -debug-info, -binary-file");
+  // correlaxtion modes(w/wo debug-info/object correlation).
```
Comment (mysterymath) [inline llvm/tools/llvm-profdata/llvm-profdata.cpp:925]: nit: I'm presuming `correlaxtion` is a typo?

**Review by mysterymath — APPROVED** [review]
LGTM for debuginfod-isms, but should probably wait for more specific feedback about `llvm-profdata` itself.

File: llvm/lib/ProfileData/InstrProfCorrelator.cpp [hunk 5]
```diff
@@ -113,11 +115,36 @@ InstrProfCorrelator::get(StringRef Filename, ProfCorrelatorKind FileKind) {
...
   if (FileKind == BINARY) {
-    auto BufferOrErr = errorOrToExpected(MemoryBuffer::getFile(Filename));
-    if (auto Err = BufferOrErr.takeError())
-      return std::move(Err);
+    if (!Filename.empty()) {
+      auto BufferOrErr = errorOrToExpected(MemoryBuffer::getFile(Filename));
+      if (auto Err = BufferOrErr.takeError())
+        return std::move(Err);
+      return get(std::move(*BufferOrErr), FileKind);
+    } else if (BIDFetcher) {
```
Comment (ellishg) [inline llvm/lib/ProfileData/InstrProfCorrelator.cpp:123]: No need for the `else` since the previous block will return.

File: llvm/lib/ProfileData/InstrProfCorrelator.cpp [hunk 6]
```diff
@@ -113,11 +115,36 @@ InstrProfCorrelator::get(StringRef Filename, ProfCorrelatorKind FileKind) {
...
+            "build IDs in a binary");
 
-    return get(std::move(*BufferOrErr), FileKind);
+      std::optional<std::string> Path = BIDFetcher->fetch(BIs->front());
+      if (Path) {
+        auto BufferOrErr = errorOrToExpected(MemoryBuffer::getFile(*Path));
+        if (auto Err = BufferOrErr.takeError())
+          return std::move(Err);
+        return get(std::move(*BufferOrErr), BINARY);
+      } else {
```
Comment (ellishg) [inline llvm/lib/ProfileData/InstrProfCorrelator.cpp:136]: If you use `if (!Path)` you can early return on the error case and remove the need to keep the above code inside an if block. I'm just trying to reduce indentation to make the code easier to read.

File: llvm/lib/ProfileData/InstrProfCorrelator.cpp [hunk 7]
```diff
@@ -113,11 +115,36 @@ InstrProfCorrelator::get(StringRef Filename, ProfCorrelatorKind FileKind) {
...
+        return make_error<InstrProfError>(
+            instrprof_error::unable_to_correlate_profile,
+            "Missing build ID: " +
+                llvm::toHex(BIs->front(), /*LowerCase=*/true));
+      }
+    } else {
```
Comment (ellishg) [inline llvm/lib/ProfileData/InstrProfCorrelator.cpp:142]: Same here. I think this can be moved up and replaced with `if (!BIDFetcher)`?

File: llvm/tools/llvm-profdata/llvm-profdata.cpp [hunk 8]
```diff
@@ -130,6 +131,12 @@ cl::opt<std::string>
                    cl::desc("For merge, use the provided unstripped bianry to "
                             "correlate the raw profile."),
                    cl::sub(MergeSubcommand));
+cl::list<std::string> DebugFileDirectory(
+    "debug-file-directory",
```
Comment (ellishg) [inline llvm/tools/llvm-profdata/llvm-profdata.cpp:135]: We already have a flag spelled `debug-info`, but this flag is unrelated.

**Review by ellishg — CHANGES_REQUESTED** [review]
I'm requesting changes because there are unresolved comments. In particular, I'm curious if you plan to support debug info correlation. One benefit of binary correlation over debug info correlation is that debug info is not required. With debuginfod, it seems you read debug info anyway.

File: llvm/tools/llvm-profdata/llvm-profdata.cpp (hunk 8 above)
Comment (petrhosek) [reply llvm/tools/llvm-profdata/llvm-profdata.cpp:135]: We already have the same option in other tools like [`llvm-cov`](https://github.com/llvm/llvm-project/blob/d8bffa9018c88ef6ce441bb44d7b7d7a9091e583/llvm/tools/llvm-cov/CodeCoverage.cpp#L684) and [`llvm-symbolizer`](https://github.com/llvm/llvm-project/blob/d8bffa9018c88ef6ce441bb44d7b7d7a9091e583/llvm/tools/llvm-symbolizer/Opts.td#L28) and I think it's worth being consistent in option spelling across tools.

File: compiler-rt/test/profile/instrprof-debuginfod-correlate.c [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+// RUN: env LLVM_PROFILE_FILE=%t.default.profraw %run %t.default.exe
+// RUN: llvm-profdata merge -o %t.default.profdata %t.default.profraw
+
+// Build with profile binary correlation and test llvm-profdata merge profile correlation with --binary-file option.
+// RUN: %clang_profgen -o %t.correlate.exe -Wl,--build-id=0x12345678 -fprofile-instr-generate -fcoverage-mapping -mllvm -profile-correlate=binary %S/Inputs/instrprof-debug-info-correlate-main.cpp %S/Inputs/instrprof-debug-info-correlate-foo.cpp
+// Strip above binary and run
+// RUN: llvm-strip %t.correlate.exe -o %t.stripped.exe
+// RUN: env LLVM_PROFILE_FILE=%t.correlate.profraw %run %t.stripped.exe
+// RUN: llvm-profdata merge -o %t.correlate-binary.profdata --binary-file=%t.correlate.exe %t.correlate.profraw
+// RUN: diff %t.default.profdata %t.correlate-binary.profdata
```
[the remark this answers is not in the document]
Comment (gulfemsavrun) [reply compiler-rt/test/profile/instrprof-debuginfod-correlate.c]: Thanks for the suggestion, I used `diff`.

File: llvm/lib/ProfileData/InstrProfCorrelator.cpp (hunk 1 above)
Comment (gulfemsavrun) [reply llvm/lib/ProfileData/InstrProfCorrelator.cpp]: The current limitation of the correlator is that it only supports correlating with a single file. When we have the support of correlating with multiple files, it should be easy to reuse that to support to correlate with multiple build IDs.

File: llvm/lib/ProfileData/InstrProfCorrelator.cpp (hunk 2 above)
Comment (gulfemsavrun) [reply llvm/lib/ProfileData/InstrProfCorrelator.cpp]: Done.

File: llvm/lib/ProfileData/InstrProfCorrelator.cpp (hunk 1 above)
Comment (gulfemsavrun) [reply llvm/lib/ProfileData/InstrProfCorrelator.cpp]: Done.

File: llvm/include/llvm/ProfileData/InstrProfReader.h (hunk 3 above)
Comment (gulfemsavrun) [reply llvm/include/llvm/ProfileData/InstrProfReader.h]: Done.

File: llvm/tools/llvm-profdata/llvm-profdata.cpp (hunk 4 above)
Comment (gulfemsavrun) [reply llvm/tools/llvm-profdata/llvm-profdata.cpp]: Done.

File: llvm/lib/ProfileData/InstrProfCorrelator.cpp (hunk 7 above)
Comment (gulfemsavrun) [reply llvm/lib/ProfileData/InstrProfCorrelator.cpp]: There is no need for that in the latest version of the patch, so I'm resolving this comment.

File: llvm/lib/ProfileData/InstrProfCorrelator.cpp (hunk 5 above)
Comment (gulfemsavrun) [reply llvm/lib/ProfileData/InstrProfCorrelator.cpp]: There is no need for that in the latest version of the patch, so I'm resolving this.

**Comment by gulfemsavrun** [comment]
> 1. Does the debuginfod correlation support work with both debug info and binary correlation? Looks like it only works with binary correlation in this change. I guess it won't be difficult to make it work with debug info correlation as well.

Thanks for the suggestion, and I extended my patch to support debuginfo correlation as well. It should work for both binary correlation and debuginfo correlation. I was just prototyping and testing it with binary correlation.

> 2. How does the new flag  `--debug-file-directory` interact with existing flag `--binary-file` which specifies just one unstripped binary?

Currently, I disallow mixing these flags. So, we can either correlate with `--debug-info`, `--binary-file` or `debuginfod` (or `--debug-file-directory`) at the moment. If we are interested in mixing them, we should have a discussion.

**Comment by gulfemsavrun** [comment]
> I'm requesting changes because there are unresolved comments. In particular, I'm curious if you plan to support debug info correlation. One benefit of binary correlation over debug info correlation is that debug info is not required. With debuginfod, it seems you read debug info anyway.

Thanks for the review @ellishg, and I added the support for debuginfo correlation.

File: llvm/tools/llvm-profdata/llvm-profdata.cpp (hunk 8 above)
Comment (gulfemsavrun) [reply llvm/tools/llvm-profdata/llvm-profdata.cpp:135]: I used the `debug-file-directory` flag to be consistent with all the other tools that supports`debuginfod`. Besides the ones that @petrhosek listed, we also have [llvm-objdump](https://github.com/llvm/llvm-project/blob/54194e1506bdd6dc37988678a8047ad4d48168fa/llvm/docs/CommandGuide/llvm-objdump.rst?plain=1#L137).

**Review by ellishg — COMMENTED** [review]
Thanks for adding support for debug info correlation! Seems ok to me but I'm curious what others think.

File: compiler-rt/test/profile/instrprof-binary-correlate-debuginfod.c [hunk 10]
```diff
@@ -0,0 +1,40 @@
+// REQUIRES: (linux || windows), curl
+// RUN: rm -rf %t
+
+// Default instrumented build with no profile correlation.
+// RUN: %clang_profgen -o %t.default.exe -Wl,--build-id=0x12345678 -fprofile-instr-generate -fcoverage-mapping %S/Inputs/instrprof-debug-info-correlate-main.cpp %S/Inputs/instrprof-debug-info-correlate-foo.cpp
+// RUN: env LLVM_PROFILE_FILE=%t.default.profraw %run %t.default.exe
+// RUN: llvm-profdata merge -o %t.default.profdata %t.default.profraw
+// RUN: llvm-profdata show %t.default.profdata > %t.default.profdata.show
```
Comment (ZequanWu) [inline compiler-rt/test/profile/instrprof-binary-correlate-debuginfod.c:8]:
```suggestion
// RUN: llvm-profdata show --all-functions --counts %t.default.profdata > %t.default.profdata.show
```
Make it consistent with the debug-info test. Plain `llvm-profdata show` doesn't show much information.

File: compiler-rt/test/profile/instrprof-binary-correlate-debuginfod.c [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+// RUN: %clang_profgen -o %t.correlate.exe -Wl,--build-id=0x12345678 -fprofile-instr-generate -fcoverage-mapping -mllvm -profile-correlate=binary %S/Inputs/instrprof-debug-info-correlate-main.cpp %S/Inputs/instrprof-debug-info-correlate-foo.cpp
+// Strip above binary and run
+// RUN: llvm-strip %t.correlate.exe -o %t.stripped.exe
+// RUN: env LLVM_PROFILE_FILE=%t.correlate.profraw %run %t.stripped.exe
+
+// Test llvm-profdata merge profile correlation with --debuginfod option.
+// RUN: mkdir -p %t/buildid/12345678
+// RUN: cp %t.correlate.exe %t/buildid/12345678/debuginfo
+// RUN: env DEBUGINFOD_CACHE_PATH=%t/debuginfod-cache DEBUGINFOD_URLS=file://%t llvm-profdata merge -o %t.correlate-debuginfod.profdata --debuginfod --correlate=binary %t.correlate.profraw
+// RUN: llvm-profdata show %t.correlate-debuginfod.profdata > %t.correlate-debuginfod.profdata.show
```
Comment (ZequanWu) [inline compiler-rt/test/profile/instrprof-binary-correlate-debuginfod.c:20]:
```suggestion
// RUN: llvm-profdata show --all-functions --counts %t.correlate-debuginfod.profdata > %t.correlate-debuginfod.profdata.show
```

File: compiler-rt/test/profile/instrprof-binary-correlate-debuginfod.c [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+// RUN: diff %t.default.profdata.show %t.correlate-debuginfod.profdata.show
+
+// Test llvm-profdata merge profile correlation with --debug-file-directory option.
+// RUN: mkdir -p %t/.build-id/12
+// RUN: cp %t.correlate.exe %t/.build-id/12/345678.debug
+// RUN: llvm-profdata merge -o %t.correlate-debug-file-dir.profdata --debug-file-directory %t --correlate=binary %t.correlate.profraw
+// RUN: llvm-profdata show %t.correlate-debug-file-dir.profdata > %t.correlate-debug-file-dir.profdata.show
```
Comment (ZequanWu) [inline compiler-rt/test/profile/instrprof-binary-correlate-debuginfod.c:27]:
```suggestion
// RUN: llvm-profdata show --all-functions --counts %t.correlate-debug-file-dir.profdata > %t.correlate-debug-file-dir.profdata.show
```

**Review by ZequanWu — APPROVED** [review]
> Currently, I disallow mixing these flags. So, we can either correlate with `--debug-info`, `--binary-file` or `debuginfod` (or `--debug-file-directory`) at the moment. If we are interested in mixing them, we should have a discussion.

Disallowing mixing these flags sounds good to me. 

LTGM, just a few nits.

File: llvm/lib/ProfileData/InstrProfCorrelator.cpp (hunk 1 above)
Comment (ZequanWu) [reply llvm/lib/ProfileData/InstrProfCorrelator.cpp]: For your reference, if you plan to add this support, I had a pending PR to add support for this feature: https://github.com/llvm/llvm-project/pull/75957.

File: compiler-rt/test/profile/instrprof-binary-correlate-debuginfod.c (hunk 10 above)
Comment (gulfemsavrun) [reply compiler-rt/test/profile/instrprof-binary-correlate-debuginfod.c]: Done.

File: compiler-rt/test/profile/instrprof-binary-correlate-debuginfod.c (hunk 12 above)
Comment (gulfemsavrun) [reply compiler-rt/test/profile/instrprof-binary-correlate-debuginfod.c]: Done.

**Comment by gulfemsavrun** [comment]
> Thanks for adding support for debug info correlation! Seems ok to me but I'm curious what others think.

@ellishg if do not have any more concerns, could you please remove the change request?

**Review by ellishg — APPROVED** [review]

**Comment by nico** [comment]
Similar to the clang discussion, or several other debuginfod discussions, I'd argue that llvm-profdata shouldn't have a dependency on an http library. If people want a way to merge local profdata for a remote executable, maybe that could be in a new binary?

**Comment by nico** [comment]
The previous discussion happened on this phab link, with https://reviews.llvm.org/D113717?id=388658#3295350 being a summary.

**Comment by petrhosek** [comment]
The approach implemented here and in all other tools that support debuginfod such as `llvm-symbolizer` is what we discussed in https://reviews.llvm.org/D113717?id=388658#3295350.
